src/analyses/cfg_dominators.h

```diff
@@ -16,17 +16,219 @@ Author: Georg Weissenbacher, georg@weissenbacher.name
+  {
+  }
+
+  dominatorst &operator=(const dominatorst &rhs)
+  {
+    dominators_data = rhs.dominators_data;
+    node_index = rhs.node_index;
+    cached_distance = rhs.cached_distance;
+    return *this;
+  }
+
+  class dominatorst_iteratort
+    : public std::iterator<std::forward_iterator_tag, T>
+  {
+  public:
+    using parentt = const datat *;
+    using elemt = const T;
+
+  private:
+    parentt data;
+    node_indext current_index;
+
+  public:
+    dominatorst_iteratort(parentt cfg_dominators, node_indext current_index)
+      : data(cfg_dominators), current_index(current_index)
+    {
+    }
+
+    dominatorst_iteratort &operator++()
+    {
+      INVARIANT(
+        current_index != npos, "Shouldn't try to increment end-iterator");
+      current_index = data->immediate_dominator[current_index];
+      return *this;
+    }
+
+    dominatorst_iteratort operator++(int)
+    {
+      INVARIANT(
+        current_index != npos, "Shouldn't try to post-increment end-iterator");
+      node_indext tmp = current_index;
+      ++(*this);
+      return dominatorst_iteratort(data, tmp);
+    }
+
+    const elemt *get() const
+    {
+      INVARIANT(
+        current_index != npos, "Shouldn't try to dereference end-iterator");
+      return &data->data[current_index];
+    }
+
+    const elemt &operator*() const
+    {
+      return *get();
+    }
+
+    const elemt *operator->() const
+    {
+      return get();
+    }
+
+    bool operator!=(const dominatorst_iteratort &other) const
+    {
+      INVARIANT(
+        data == other.data, "iterators from different sets are not comparable");
+      return current_index != other.current_index;
+    }
+
+    bool operator==(const dominatorst_iteratort &other) const
+    {
+      return !(*this != other);
+    }
+  };
+
+  using const_iterator = dominatorst_iteratort;
+
+  const_iterator begin() const
+  {
+    return const_iterator(dominators_data.get(), node_index);
+  }
+
+  const_iterator cbegin() const
+  {
+    return begin();
+  }
+
+  const_iterator end() const
+  {
+    return const_iterator(dominators_data.get(), npos);
+  }
+
+  const_iterator cend() const
+  {
+    return end();
+  }
+
+  /// Return an iterator node if node is in this dominators set, end() otherwise
+  /// Note: O(n), when making many queries against the same set it is probably
+  /// worth copying into a std::set
+  const_iterator find(const T &node) const
+  {
+    std::less<T> less;
+    // FIXME This works around a bug in other parts of the code in particular,
+    // dependence_graph.cpp, where iterators to different lists than those that
+    // are stored in this set are passed to find. The Debug libstdc++ will
+    // (correctly!) run into an assertion failure using std::find. std::less for
+    // some reason doesn't trigger this assertion failure, so we use this as an
+    // ugly workaround until that code is fixed.
+
+    // NOLINTNEXTLINE
+    return std::find_if(cbegin(), cend(), [&](const T &other_node) {
```

Comment:
How about just comparing by location number? Those should be unique (if not, the caller must run `goto_programt::update()` first)

Or, for other T, require that the supplied T provide some accessor that yields a unique ID.

Reply:
That might work, I don't think I was aware of location numbers at the time or that they were meant to be unique.